{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Abstract\n",
    "\n",
    " Word2Vec을 활용하여 연도별로 GNI 코퍼스 내의 keywords들을 Clusturing한다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import nltk\n",
    "import nltk\n",
    "from nltk.corpus import *\n",
    "from nltk import *\n",
    "\n",
    "# libraries for word clustering\n",
    "from gensim.models import Word2Vec\n",
    "from sklearn.cluster import KMeans\n",
    "import joblib\n",
    "\n",
    "# import libraries for preprocessing\n",
    "import re\n",
    "\n",
    "# libraries for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from sklearn.manifold import MDS\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using sentence tokenized gni corpus.\n",
    "corpus_root = 'C:/Users/Maeg/nltk_data/Genomics-Informatics-Corpus-master/Genomics-Informatics-Corpus-master/GNI Corpus 1.0/sentence_tokenized'\n",
    "GNIcorpus = PlaintextCorpusReader(corpus_root,'.*\\.txt',encoding='utf-8')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 전처리"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "# preprocessing each sentence\n",
    "def gni_preprocessed_sents(gni_sents):\n",
    "    \n",
    "    # load stopwords and initialize preprocessed sentences list\n",
    "    stops = set(stopwords.words(\"english\"))\n",
    "    wnl = nltk.WordNetLemmatizer()\n",
    "    gni_pre_sents = []\n",
    "    \n",
    "    # remove non-letter words and stop words in each sentence\n",
    "    for s in gni_sents :\n",
    "        sent = []\n",
    "        for w in s :\n",
    "            if (re.match(r'[a-zA-Z0-9]+',w) !=  None) & (not w in stops) :\n",
    "                sent.append(wnl.lemmatize(w.lower()))\n",
    "        gni_pre_sents.append(sent)\n",
    "    \n",
    "    # return preprocessed sentences list\n",
    "    return gni_pre_sents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start preprocessing... : G&I sentences\n",
      "End preprocessing: G&I sentences\n"
     ]
    }
   ],
   "source": [
    "# preprocessing GNI corpus\n",
    "print(\"Start preprocessing... : G&I sentences\")\n",
    "gni_sents = gni_preprocessed_sents(GNIcorpus.sents())\n",
    "print(\"End preprocessing: G&I sentences\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modeling and Training Word2Vec "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start Training ...\n",
      "End Training \n"
     ]
    }
   ],
   "source": [
    "n_features = 300       # the number of Word2Vector Dimension\n",
    "min_wordcount = 40     # Words which counts at least 20 times are available\n",
    "n_workers = 4          # the number of parallel work\n",
    "size_boundary = 10      # contextual window size\n",
    "downsampling = 1e-3    # down sampling with frequent words\n",
    "\n",
    "# Initialize and train Word2Vec model\n",
    "print(\"Start Training ...\")\n",
    "model = Word2Vec(gni_sents,\n",
    "                 workers = n_workers,\n",
    "                 size = n_features,\n",
    "                 min_count = min_wordcount,\n",
    "                 window = size_boundary,\n",
    "                 sample = downsampling)\n",
    "print(\"End Training \")\n",
    "\n",
    "# manage memory\n",
    "model.init_sims(replace=True)\n",
    "\n",
    "# save trained model\n",
    "model.save(\"GNI_\"+str(n_features)+\"features_\"+str(min_wordcount)+\"minwords_\"+str(size_boundary)+\"window\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Word Clustering\n",
    "\n",
    "Word2Vec으로 얻은 각 단어의 word vector를 이용하여 word cluster한다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start: Clustering\n",
      "End  : Clustering\n",
      "\n",
      "============== Clustering Information ==============\n",
      "Clustering time: 38.91790795326233\n",
      "# of Word vectors: 2191\n",
      "# of Clusters: 438\n"
     ]
    }
   ],
   "source": [
    "# clustering start time\n",
    "start = time.time()\n",
    "print(\"Start: Clustering\")\n",
    "\n",
    "# word vector of word2Vec model\n",
    "gniwv = model.wv\n",
    "\n",
    "# set the number of cluster\n",
    "num_clusters = (int)(len(gniwv.vocab) / 5)\n",
    "\n",
    "# Initialize a k-means object and use it to extract centroids(cluster)\n",
    "kmeans_clustering = KMeans( n_clusters = num_clusters)\n",
    "index = kmeans_clustering.fit_predict(gniwv.vectors)\n",
    "\n",
    "end = time.time()\n",
    "clustering_time = end - start\n",
    "print(\"End  : Clustering\")\n",
    "\n",
    "# print Clustering information\n",
    "print(\"\\n============== Clustering Information ==============\")\n",
    "print(\"Clustering time:\", clustering_time)\n",
    "print(\"# of Word vectors:\", len(gniwv.vocab))\n",
    "print(\"# of Clusters:\", num_clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GNI_words_cluster.pkl']"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save the cluster models\n",
    "joblib.dump(kmeans_clustering, 'GNI_words_cluster.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Cluster 0\n",
      "['still', 'microbial', 'limited', 'advance', 'epigenome', 'effort', 'metagenomic', 'technical', 'enabled']\n",
      "\n",
      "Cluster 1\n",
      "['at', 'degree', 'similarly', 'circle', 'fish', 'black']\n",
      "\n",
      "Cluster 2\n",
      "['altered', 'oncogenic']\n",
      "\n",
      "Cluster 3\n",
      "['standard', 'initial', 'obtain', 'check', 'normalization', 'metric']\n",
      "\n",
      "Cluster 4\n",
      "['obtained', 'quality', 'count', 'after', 'call', 'removed', 'filtered']\n",
      "\n",
      "Cluster 5\n",
      "['population', 'asian', 'breed', 'cattle', 'hanwoo', 'holstein', 'african', 'pig', 'ancestry', 'east']\n",
      "\n",
      "Cluster 6\n",
      "['occurred', 'cow', 'period', 'metritis', 'prevalent', 'half']\n",
      "\n",
      "Cluster 7\n",
      "['faire', 'tss', 'upstream', 'island', 'transcribed', 'h3k4me3', 'h3k9', 'nucleosome', 'h3k4me1']\n",
      "\n",
      "Cluster 8\n",
      "['20', '30', '50', '40', '60', '80', 'mm', '70', '65']\n",
      "\n",
      "Cluster 9\n",
      "['approach', 'developed', 'knowledge', 'make', 'technique', 'computational', 'mining', 'efficient']\n",
      "\n",
      "Cluster 10\n",
      "['increase', 'negative', 'reduced', 'decrease', 'specificity']\n",
      "\n",
      "Cluster 11\n",
      "['search', 'query', 'blast']\n",
      "\n",
      "Cluster 12\n",
      "['cell', 'mouse', 'liver', 'brain', 'carcinoma', 'death', 'heart', 'prostate', 'colon', 'lymphoma', 'renal', 'adenocarcinoma', 'pancreatic']\n",
      "\n",
      "Cluster 13\n",
      "['proportion', 'substitution', 'explained', 'assumed']\n",
      "\n",
      "Cluster 14\n",
      "['most', 'exist', 'newly', 'indeed', 'substantial', 'extent', 'accumulation']\n",
      "\n",
      "Cluster 15\n",
      "['needed', 'developing', 'appropriate', 'trial', 'personalized', 'benefit', 'care', 'cma']\n",
      "\n",
      "Cluster 16\n",
      "['e6', 'na', 'ns3', 'zikv', 'ofloxacin', 'lpxd', 'curcumin', 'antagonist', 'agonist', 'rif']\n",
      "\n",
      "Cluster 17\n",
      "['tested']\n",
      "\n",
      "Cluster 18\n",
      "['according', 'protocol', 'fisher', 'japan', 'city']\n",
      "\n",
      "Cluster 19\n",
      "['directly', 'produce', 'editing', 'restriction', 'isolation']\n",
      "\n",
      "Cluster 20\n",
      "['comparison', 'each', 'chosen', 'display', 'clustered', 'heatmap', 'displayed', 'drawn', 'visualized']\n",
      "\n",
      "Cluster 21\n",
      "['whether', 'often', 'impact', 'explain', 'causal']\n",
      "\n",
      "Cluster 22\n",
      "['regulated', 'pkd2', 'upregulated', 'pten', 'downregulated', 'mef']\n",
      "\n",
      "Cluster 23\n",
      "['project', 'resource', 'published', 'center', 'institute', 'international', 'consortium']\n",
      "\n",
      "Cluster 24\n",
      "['includes', 'mesh', 'portal', 'mysql', 'keywords', 'ictv']\n",
      "\n",
      "Cluster 25\n",
      "['age', 'woman', 'crc', 'child', 'kd', 'gastritis', 'month', 'prevalence', 'incidence', 'sle', 'younger', 'ht']\n",
      "\n",
      "Cluster 26\n",
      "['produced', 'phase', 'pre', 'preparation', 'prepared', 'bio', 'processed']\n",
      "\n",
      "Cluster 27\n",
      "['relation', 'virulence', 'baumannii', 'acinetobactin', 'trans']\n",
      "\n",
      "Cluster 28\n",
      "['mean', 'follows', 'confidence', 'log', 'maximum', 'deviation', 'square', 'minimum', 'quantile', 'q', 'mig']\n",
      "\n",
      "Cluster 29\n",
      "['akt2', 'akt3']\n",
      "\n",
      "Cluster 30\n",
      "['frequent']\n",
      "\n",
      "Cluster 31\n",
      "['1', '2', '3', '4']\n",
      "\n",
      "Cluster 32\n",
      "['mechanism', 'potential', 'epigenetic', 'key', 'critical', 'architecture']\n",
      "\n",
      "Cluster 33\n",
      "['commonly', 'note', 'confirm', 'ass', 'little', 'clearly', 'unlike', 'situation']\n",
      "\n",
      "Cluster 34\n",
      "['bacteria', 'organism', 're', 'eukaryote']\n",
      "\n",
      "Cluster 35\n",
      "['genomic', 'amount']\n",
      "\n",
      "Cluster 36\n",
      "['sva', 'contain', 'around', 'retrotransposons', 'l1', 'ltr', 'duplication', 'segment', 'inserted']\n",
      "\n",
      "Cluster 37\n",
      "['complete', 'construction', 'workflow', 'ivag']\n",
      "\n",
      "Cluster 38\n",
      "['simple', 'basic', 'pubannotation', 'taxonomy', 'interface', 'token', 'grid', 'framework', 'document', 'inception', 'created', 'layer', 'patent', 'lapp', 'object', 'schema', 'abstract', 'create', 'automatically', 'nlp']\n",
      "\n",
      "Cluster 39\n",
      "['http', 'www', 'org', 'edu', 'com', 'ac']\n",
      "\n",
      "Cluster 40\n",
      "['region', 'chromosome', 'intron', 'located', 'boundary', 'flanking', 'overlapped', 'intergenic', 'intronic', 'genic']\n",
      "\n",
      "Cluster 41\n",
      "['expressed', 'differentially', 'hypermethylation']\n",
      "\n",
      "Cluster 42\n",
      "['best', 'binary', 'blup', 'covariance']\n",
      "\n",
      "Cluster 43\n",
      "['survival', 'hepatitis', 'adult', 'oral', 'vpa', 'adrs']\n",
      "\n",
      "Cluster 44\n",
      "['insulin', 'lipid', 'strongly', 'correlated', 'lipoprotein']\n",
      "\n",
      "Cluster 45\n",
      "['donor', 'isolated', 'cfdna', 'urine', 'peripheral']\n",
      "\n",
      "Cluster 46\n",
      "['statistical', 'performance', 'accuracy', 'procedure', 'mdr', 'predictive']\n",
      "\n",
      "Cluster 47\n",
      "['estimated', 'missing', 'r2', 'equilibrium', 'yorkshire', 'landrace', 'hardy', 'weinberg']\n",
      "\n",
      "Cluster 48\n",
      "['tree', 'vector', 'us', 'markov', 'suffix']\n",
      "\n",
      "Cluster 49\n",
      "['section', 'last', 'kim', 'reviewed', 'lee', 'follow', 'united']\n",
      "\n",
      "Cluster 50\n",
      "['de', 'assembly', 'novo', 'rice']\n",
      "\n",
      "Cluster 51\n",
      "['predicted', 'ion', '3d', 'backbone']\n",
      "\n",
      "Cluster 52\n",
      "['this', 'therefore', 'understanding', 'help', 'understand', 'insight', 'relevant', 'providing']\n",
      "\n",
      "Cluster 53\n",
      "['sequencing', 'technology', 'ng', 'next', 'generation', 'exome', 'throughput']\n",
      "\n",
      "Cluster 54\n",
      "['28', '29', '33', '37', '36', '48', '52', '66']\n",
      "\n",
      "Cluster 55\n",
      "['though', 'far', 'through', 'date', 'systematic', 'tried', 'fundamental', 'nevertheless', 'toward', 'completely']\n",
      "\n",
      "Cluster 56\n",
      "['containing', 'presence', 'loop', 'terminal', 'superfamily', 'helix', 'transition', 'alpha', 'bound']\n",
      "\n",
      "Cluster 57\n",
      "['investigated', 'gwass']\n",
      "\n",
      "Cluster 58\n",
      "['file', 'alignment', 'map', 'format', 'input', 'output', 'build']\n",
      "\n",
      "Cluster 59\n",
      "['transcription', 'chromatin', 'modification', 'histone']\n",
      "\n",
      "Cluster 60\n",
      "['function', 'well', 'process', 'cellular', 'essential']\n",
      "\n",
      "Cluster 61\n",
      "['inhibitor', 'cox', 'migration', 'inflammatory', 'involvement', 'vitro', 'rage', 'secretion', 'vivo', 'senescence']\n",
      "\n",
      "Cluster 62\n",
      "['f']\n",
      "\n",
      "Cluster 63\n",
      "['shown', '1b', '1a', '2a', '2b', '3a']\n",
      "\n",
      "Cluster 64\n",
      "['interaction', 'biological', 'network']\n",
      "\n",
      "Cluster 65\n",
      "['animal', 'world', 'past', 'country']\n",
      "\n",
      "Cluster 66\n",
      "['read', 'length', 'base', 'mapped', 'raw', 'peak', 'bed', 'aligned']\n",
      "\n",
      "Cluster 67\n",
      "['to', 'multiple', 'detect', 'reliable', 'detecting', 'analyzing']\n",
      "\n",
      "Cluster 68\n",
      "['generated', 'generate', 'series', 'capture', 'digital', 'track']\n",
      "\n",
      "Cluster 69\n",
      "['if', 'optimal', 'consider', 'hypothesis', 'define', 'generalized']\n",
      "\n",
      "Cluster 70\n",
      "['n', 'x', 'd', 's', 'v', 'l', 'j', 'y', 'upper']\n",
      "\n",
      "Cluster 71\n",
      "['include', 'soil', 'ability', 'nature', 'recognized', 'subsequently', 'during', 'accessibility', 'involving', 'discussed', 'capable', 'microorganism']\n",
      "\n",
      "Cluster 72\n",
      "['independent', 'assumption', 'ethnicity', 'gwa']\n",
      "\n",
      "Cluster 73\n",
      "['usually', 'chromosomal', 'mark', 'indicating', 'noncoding', 'spatial']\n",
      "\n",
      "Cluster 74\n",
      "['except', '61', '67', '73', '89', '86', '92', '78', '88']\n",
      "\n",
      "Cluster 75\n",
      "['clinical', 'biomarkers', 'heterogeneity', 'prognostic']\n",
      "\n",
      "Cluster 76\n",
      "['another', 'design', 'focus', 'regarding', 'develop', 'link', 'regard', 'explore']\n",
      "\n",
      "Cluster 77\n",
      "['side', 'amplified', 'sense', 'gc', '150', 'mer', 'barcodes']\n",
      "\n",
      "Cluster 78\n",
      "['consists', 'composed', 'diagram', 'connected']\n",
      "\n",
      "Cluster 79\n",
      "['positive', 'overall', 'false', 'true', 'absolute']\n",
      "\n",
      "Cluster 80\n",
      "['distribution', 'ratio', 'fold', 'percentage', 'axis']\n",
      "\n",
      "Cluster 81\n",
      "['tool', 'annotation', 'available', 'user', 'provides', 'integrated', 'visualization']\n",
      "\n",
      "Cluster 82\n",
      "['they', 'frequently', 'occur', 'rearrangement']\n",
      "\n",
      "Cluster 83\n",
      "['element', 'long', 'alu', 'event', 'te']\n",
      "\n",
      "Cluster 84\n",
      "['mutation', 'somatic']\n",
      "\n",
      "Cluster 85\n",
      "['end', 'fragment', 'primer', 'strand', 'cdna', 'stranded', 'dnase']\n",
      "\n",
      "Cluster 86\n",
      "['gwas', 'dataset', 'all', 'kare', 'combined', 'genomewide', 'summary', 'catalog']\n",
      "\n",
      "Cluster 87\n",
      "['detected', 'examined', 'confirmed', 'consistently']\n",
      "\n",
      "Cluster 88\n",
      "['location', 'portion', 'carry']\n",
      "\n",
      "Cluster 89\n",
      "['resulting', 'manner', 'stable', 'intermediate', 'exhibited', 'nine', 'light', 'endogenous']\n",
      "\n",
      "Cluster 90\n",
      "['block', 'haplotype', 'ld', 'linkage', 'disequilibrium']\n",
      "\n",
      "Cluster 91\n",
      "['there', 'focused', 'yet', 'importance', 'investigation', 'underlying', 'contribution', 'aimed', 'implication']\n",
      "\n",
      "Cluster 92\n",
      "['cause', 'loss', 'disorder', 'leading', 'caused', 'suggesting']\n",
      "\n",
      "Cluster 93\n",
      "['bmi', 'density', 'mass', 'grs', 'bmr', 'statistically', 'null', 'tt', 'gstm1']\n",
      "\n",
      "Cluster 94\n",
      "['certain', 'depending', 'origin', 'clear', 'largely', 'selective', 'closely', 'inter', 'neuronal', 'hbv', 'morphology', 'severe', 'ifi6', 'defect']\n",
      "\n",
      "Cluster 95\n",
      "['binding', 'active', 'bind', 'affinity', 'interact', 'substrate', 'hydrophobic', 'ama1', 'finger']\n",
      "\n",
      "Cluster 96\n",
      "['entity', 'task', 'existing', 'named', 'utilized', 'neural', 'ner']\n",
      "\n",
      "Cluster 97\n",
      "['database', 'annotated', 'ncbi', 'downloaded', 'accession', 'genbank']\n",
      "\n",
      "Cluster 98\n",
      "['5', '10', '6', '8', '7']\n",
      "\n",
      "Cluster 99\n",
      "['gi', 'release', '2013', 'website', 'orcid', 'released']\n",
      "\n",
      "Cluster 100\n",
      "['evaluated', 'sensitivity', 'permutation', 'standardized', 'adjusted']\n",
      "\n",
      "Cluster 101\n",
      "['diagnosis', 'prevention']\n",
      "\n",
      "Cluster 102\n",
      "['a']\n",
      "\n",
      "Cluster 103\n",
      "['two', 'three']\n",
      "\n",
      "Cluster 104\n",
      "['following', 'run', 'plink', 'final', 'o', 'gb', 'default']\n",
      "\n",
      "Cluster 105\n",
      "['history', 'outcome', 'gdm', 'facial', 'medication', 'conventional', 't2dm', 'adherence']\n",
      "\n",
      "Cluster 106\n",
      "['showed', 'observed', 'strong', 'showing']\n",
      "\n",
      "Cluster 107\n",
      "['variant', 'variation', 'rare']\n",
      "\n",
      "Cluster 108\n",
      "['mirnas', 'regulation', 'mrna', 'revealed', 'mirna', 'transcriptional', 'trf', 'lncrnas', 'silencing', 'nc886', 'aberrant']\n",
      "\n",
      "Cluster 109\n",
      "['pair', 'distance', 'least', 'pairwise', 'zero']\n",
      "\n",
      "Cluster 110\n",
      "['relative', 'abundance']\n",
      "\n",
      "Cluster 111\n",
      "['considering', 'additionally', 'reduction', 'rather', 'canonical', 'other', 'but', 'share', 'strength', 'examine', 'continuous', 'infer']\n",
      "\n",
      "Cluster 112\n",
      "['usa', 'ca', 'real', 'right', 'inc', 'manufacturer', 'volume', 'instruction', 'qiagen', 'mixture', 'ma', 'san', 'plate', 'diego']\n",
      "\n",
      "Cluster 113\n",
      "['research', 'genomics', 'issue', 'medical', 'medicine', 'life', 'scientific', 'science', 'informatics', 'genetics', 'management', 'fda']\n",
      "\n",
      "Cluster 114\n",
      "['i']\n",
      "\n",
      "Cluster 115\n",
      "['furthermore', 'mainly', 'able', 'suggests', 'signature', 'some', 'heterogeneous', 'possibility', 'driver', 'proven', 'phenomenon', 'potentially', 'serve', 'consequence']\n",
      "\n",
      "Cluster 116\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['addition', 'contrast']\n",
      "\n",
      "Cluster 117\n",
      "['repeat', 'motif', 'start', 'homologous', 'frame', 'reading', 'tandem', 'sine']\n",
      "\n",
      "Cluster 118\n",
      "['similarity', 'cluster', 'node', 'represented', 'edge']\n",
      "\n",
      "Cluster 119\n",
      "['human', 'specie', 'plant']\n",
      "\n",
      "Cluster 120\n",
      "['sample']\n",
      "\n",
      "Cluster 121\n",
      "['putative', 'zinc', 'interacting', 'duplex']\n",
      "\n",
      "Cluster 122\n",
      "['physical', 'smoking', 'alcohol', 'consumption', 'intake']\n",
      "\n",
      "Cluster 123\n",
      "['present', 'particular', 'fact', 'conclusion']\n",
      "\n",
      "Cluster 124\n",
      "['involved', 'mediated', 'activation', 'metabolism', 'formation', 'act', 'regulate', 'targeting', 'emt', 'tm4sf5', 'regulator', 'stress', 'transport', 'regulating', 'biosynthesis', 'initiation', 'transduction']\n",
      "\n",
      "Cluster 125\n",
      "['on', 'indicates', 'hand', 'tc', 'cai']\n",
      "\n",
      "Cluster 126\n",
      "['respectively', 'ci', '64', '51', 'mg', '57', '75', '85', 'dl', '83', '71', '79']\n",
      "\n",
      "Cluster 127\n",
      "['area', 'curve', 'auc', 'roc', 'operating', 'receiver']\n",
      "\n",
      "Cluster 128\n",
      "['data', 'datasets']\n",
      "\n",
      "Cluster 129\n",
      "['patient', 'normal', 'blood', 'plasma']\n",
      "\n",
      "Cluster 130\n",
      "['title']\n",
      "\n",
      "Cluster 131\n",
      "['expected', 'much', 'relatively']\n",
      "\n",
      "Cluster 132\n",
      "['genetic', 'trait', 'phenotype', 'candidate']\n",
      "\n",
      "Cluster 133\n",
      "['bp', 'state', '200', 'unit']\n",
      "\n",
      "Cluster 134\n",
      "['regression', 'logistic', 'lasso', 'covariates', 'pca', 'additive']\n",
      "\n",
      "Cluster 135\n",
      "['modern', 'neanderthal']\n",
      "\n",
      "Cluster 136\n",
      "['instead', 'prior', 'scheme', 'setting', 'remove', 'optimization', 'internal', 'typically', 'fusionscan', 'character', 'typical', 'assessed', 'po', 'verified', 'reliability']\n",
      "\n",
      "Cluster 137\n",
      "['0', 'p', '05', '01', 'height', '04', 'odds', 'fdr', '001', 'cutoff', 'counted']\n",
      "\n",
      "Cluster 138\n",
      "['future', 'established', 'led', 'progress', 'promising']\n",
      "\n",
      "Cluster 139\n",
      "['value', 'plot', 'calculated', 'coefficient', 'threshold']\n",
      "\n",
      "Cluster 140\n",
      "['simulation', 'pharmacophore', 'screen', 'md']\n",
      "\n",
      "Cluster 141\n",
      "['domain', 'conserved']\n",
      "\n",
      "Cluster 142\n",
      "['example']\n",
      "\n",
      "Cluster 143\n",
      "['korea', 'university', 'health', 'national', 'seoul', 'dr', 'hospital']\n",
      "\n",
      "Cluster 144\n",
      "['duplicate', '2000', 'limit', 'qc', 'mgiseq']\n",
      "\n",
      "Cluster 145\n",
      "['b', 'c', 't']\n",
      "\n",
      "Cluster 146\n",
      "['current', 'conducted', 'precision', 'approved', 'written', 'biobanks', 'consent', 'improvement', 'laboratory', 'board', 'informed']\n",
      "\n",
      "Cluster 147\n",
      "['disease', 'related', 'metabolic', 'hypertension']\n",
      "\n",
      "Cluster 148\n",
      "['18', '17', '25', '23', '19', '22', '24', '26', '27']\n",
      "\n",
      "Cluster 149\n",
      "['single', 'polymorphism', 'nucleotide']\n",
      "\n",
      "Cluster 150\n",
      "['production', 'characterized', 'specifically', 'resulted', 'induce', 'enhanced', 'induction', 'transporter', 'implicated', 'cytokine', 'induces']\n",
      "\n",
      "Cluster 151\n",
      "['al']\n",
      "\n",
      "Cluster 152\n",
      "['pcr', 'reaction', 'amplification', 'assay', 'kit']\n",
      "\n",
      "Cluster 153\n",
      "['localization', 'stability', 'folding']\n",
      "\n",
      "Cluster 154\n",
      "['minor', 'maf', 'excluded', 'platelet']\n",
      "\n",
      "Cluster 155\n",
      "['mutant', 'indicated', 'wild', 'mrp2', '1β', '74', 'glut2']\n",
      "\n",
      "Cluster 156\n",
      "['specific', 'pattern']\n",
      "\n",
      "Cluster 157\n",
      "['natural']\n",
      "\n",
      "Cluster 158\n",
      "['array', 'affymetrix', 'santa', 'clara', '1000']\n",
      "\n",
      "Cluster 159\n",
      "['composition']\n",
      "\n",
      "Cluster 160\n",
      "['depth', 'illumina', 'coverage', 'paired']\n",
      "\n",
      "Cluster 161\n",
      "['chain', 'id', 'template', 'pdb', 'kr']\n",
      "\n",
      "Cluster 162\n",
      "['cancer', 'tumor', 'tissue', 'lung', 'breast', 'primary', 'kidney', 'colorectal', 'ovarian']\n",
      "\n",
      "Cluster 163\n",
      "['90', 'interval', 'remaining', 'every', '99', 'fraction', 'ranging', '98', 'sd', 'chr1', '93', 'ranged', '400']\n",
      "\n",
      "Cluster 164\n",
      "['may', 'it', 'could', 'important', 'thus', 'these', 'might']\n",
      "\n",
      "Cluster 165\n",
      "['whereas', '35', '45', '34', '42', '41', '55', '53', '54', '58', '68', '59', '62']\n",
      "\n",
      "Cluster 166\n",
      "['subsequent', 'extraction', 'deep', 'quantification', 'metagenome', 'initially', 'massive', 'parallel', 'availability', 'hybrid']\n",
      "\n",
      "Cluster 167\n",
      "['translation', 'translational', 'post', 'transfer', 'thereby', 'ncrnas', 'isoforms', 'machinery', 'lncrna']\n",
      "\n",
      "Cluster 168\n",
      "['indicate', 'when', 'dominant', 'almost', 'phenotypic', 'close', 'tend']\n",
      "\n",
      "Cluster 169\n",
      "['significantly', 'tg', 'decreased', 'fat', 'hdlc', 'cholesterol', 'hdl']\n",
      "\n",
      "Cluster 170\n",
      "['associated', 'common', 'susceptibility']\n",
      "\n",
      "Cluster 171\n",
      "['first', 'original']\n",
      "\n",
      "Cluster 172\n",
      "['using', 'performed']\n",
      "\n",
      "Cluster 173\n",
      "['korean', 'subject', 'included', 'sex', 'heritability', 't1dm', 'fulminant', 'athlete']\n",
      "\n",
      "Cluster 174\n",
      "['ago', 'let', '196b', 'phosphate', 'unigenes']\n",
      "\n",
      "Cluster 175\n",
      "['code', 'along', 'seed', 'contained', 'master']\n",
      "\n",
      "Cluster 176\n",
      "['snp', 'marker']\n",
      "\n",
      "Cluster 177\n",
      "['thousand', 'hundred', 'ten']\n",
      "\n",
      "Cluster 178\n",
      "['together', 'view', 'while', 'integration', 'compositional', 'integrating', 'integrative', 'proteomic']\n",
      "\n",
      "Cluster 179\n",
      "['hpv', 'petrii', 'protease', 'namely', 'iron', 'surface', 'influenza', 'cysq', 'catalytic', 'coupled']\n",
      "\n",
      "Cluster 180\n",
      "['type']\n",
      "\n",
      "Cluster 181\n",
      "['total', 'average', '100', 'per']\n",
      "\n",
      "Cluster 182\n",
      "['however', 'due', 'since', 'even', 'diversity']\n",
      "\n",
      "Cluster 183\n",
      "['cpg', 'methylated', 'cgis']\n",
      "\n",
      "Cluster 184\n",
      "['given', 'measure']\n",
      "\n",
      "Cluster 185\n",
      "['differential', 'degs']\n",
      "\n",
      "Cluster 186\n",
      "['general', 'detailed', 'shared', 'composite', 'overview', 'store']\n",
      "\n",
      "Cluster 187\n",
      "['divided', 'randomly']\n",
      "\n",
      "Cluster 188\n",
      "['new', 'application', 'review', 'researcher', 'work', 'strategy', 'field', 'community', 'biology', 'challenge']\n",
      "\n",
      "Cluster 189\n",
      "['modified', 'added', 'rule', 'hmm', 'hidden', 'building', 'label', 'entry', 'descriptor', 'refined']\n",
      "\n",
      "Cluster 190\n",
      "['seq', 'microarray', 'transcriptome', 'profiling']\n",
      "\n",
      "Cluster 191\n",
      "['acid', 'amino']\n",
      "\n",
      "Cluster 192\n",
      "['class']\n",
      "\n",
      "Cluster 193\n",
      "['second', 'validation', 'cross', 'done', 'calculation', 'compare', 'definition', 'achieved', 'recommended', 'bayesian', 'robust', 'calculate', 'multivariate']\n",
      "\n",
      "Cluster 194\n",
      "['of', 'milk', 'autosomal', 'nonsynonymous']\n",
      "\n",
      "Cluster 195\n",
      "['from', 'subset', 'representative', 'seen', 'quite', 'taxon', 'figure']\n",
      "\n",
      "Cluster 196\n",
      "['word', 'an', 'take', 'proper', 'vocabulary', 'inference']\n",
      "\n",
      "Cluster 197\n",
      "['nimbolide', 'ir', 'muscle', 'h460r', 'parental', 'sntb2', 'rat']\n",
      "\n",
      "Cluster 198\n",
      "['strain', 'coli', 'rapa', 'japonicum']\n",
      "\n",
      "Cluster 199\n",
      "['encode', 'open', 'encyclopedia']\n",
      "\n",
      "Cluster 200\n",
      "['employed', 'fast', 'solution', 'storage', 'iabio', 'dependency', 'visual', 'analytical', 'description', 'hiscom', 'external', 'image', 'interactive']\n",
      "\n",
      "Cluster 201\n",
      "['step', 'platform', 'applied', 'pipeline', 'experiment', 'option', 'perform']\n",
      "\n",
      "Cluster 202\n",
      "['stage', 'gastric', 'copd', 'bone']\n",
      "\n",
      "Cluster 203\n",
      "['part', 'detail', 'ihec', 'cover', 'journal']\n",
      "\n",
      "Cluster 204\n",
      "['tag', 'probe', 'bar', 'intensity', 'normalized', 'identity', 'root']\n",
      "\n",
      "Cluster 205\n",
      "['form', 'double', 'transposition', 'gyrase', 'tail', 'cleavage']\n",
      "\n",
      "Cluster 206\n",
      "['mostly', 'occurs', 'overlapping', 'overlap', 'direction', 'marked', 'cpgis', 'producing']\n",
      "\n",
      "Cluster 207\n",
      "['variety', 'global', 'proteome']\n",
      "\n",
      "Cluster 208\n",
      "['term', 'go', 'ontology']\n",
      "\n",
      "Cluster 209\n",
      "['recent', 'report']\n",
      "\n",
      "Cluster 210\n",
      "['structure', 'property', 'structural', 'hp', 'secondary', 'homology']\n",
      "\n",
      "Cluster 211\n",
      "['exon', 'deletion', 'insertion']\n",
      "\n",
      "Cluster 212\n",
      "['group']\n",
      "\n",
      "Cluster 213\n",
      "['carried', 'modeling', 'dimensional', 'space', 'representation', 'optimized', 'subjected', 'graphical']\n",
      "\n",
      "Cluster 214\n",
      "['our']\n",
      "\n",
      "Cluster 215\n",
      "['use', 'problem', 'made', 'necessary', 'limitation', 'with', 'accurate', 'advantage', 'allows', 'overcome', 'practice']\n",
      "\n",
      "Cluster 216\n",
      "['gene', 'enrichment']\n",
      "\n",
      "Cluster 217\n",
      "['smaller', 'yield', 'exact', 'heterozygosity', 'heterozygous', 'allelic', 'concordance', 'outlier']\n",
      "\n",
      "Cluster 218\n",
      "['response', 'growth', 'induced', 'anti', 'apoptosis', 'proliferation', 'suppressor', 'differentiation', 'invasion', 'abnormal', 'epithelial', 'oncogene', 'overexpression', 'tmprss4', 'inflammation']\n",
      "\n",
      "Cluster 219\n",
      "['target']\n",
      "\n",
      "Cluster 220\n",
      "['reported', 'previously']\n",
      "\n",
      "Cluster 221\n",
      "['ne', 'people', 'recruited', 'south', 'american']\n",
      "\n",
      "Cluster 222\n",
      "['others', 'cad', 'aspirin', 'ra']\n",
      "\n",
      "Cluster 223\n",
      "['entire', 'variability', 'allowed', 'gap', 'starting', 'blat']\n",
      "\n",
      "Cluster 224\n",
      "['including', 'condition', 'progression', 'immune', 'metastasis', 'developmental', 'agent']\n",
      "\n",
      "Cluster 225\n",
      "['fig', 'table', 'supplementary']\n",
      "\n",
      "Cluster 226\n",
      "['defined', 'k', 'represent', 'represents']\n",
      "\n",
      "Cluster 227\n",
      "['one']\n",
      "\n",
      "Cluster 228\n",
      "['literature', 'record', 'kimchi']\n",
      "\n",
      "Cluster 229\n",
      "['fluid', 'distinguish', 'transplant', 'biopsy', 'pdx', 'dpcr']\n",
      "\n",
      "Cluster 230\n",
      "['frequency', 'allele', 'hla']\n",
      "\n",
      "Cluster 231\n",
      "['large', 'whole', 'scale']\n",
      "\n",
      "Cluster 232\n",
      "['testing', 'admixture', 'dtc', 'consumer', 'company']\n",
      "\n",
      "Cluster 233\n",
      "['significant', 'correlation', 'significance']\n",
      "\n",
      "Cluster 234\n",
      "['described', 'presented']\n",
      "\n",
      "Cluster 235\n",
      "['high', 'low']\n",
      "\n",
      "Cluster 236\n",
      "['test', 'estimate', 'power']\n",
      "\n",
      "Cluster 237\n",
      "['e', 'g']\n",
      "\n",
      "Cluster 238\n",
      "['bmd', 'lpl', 'mets', 'klf5', 'lifestyle', 'lef1']\n",
      "\n",
      "Cluster 239\n",
      "['among']\n",
      "\n",
      "Cluster 240\n",
      "['library', 'chip', 'resolution', 'hybridization']\n",
      "\n",
      "Cluster 241\n",
      "['experimental', 'widely', 'multi', 'here', 'analyze', 'purpose', 'interpretation', 'goal', 'aim', 'powerful', 'easy', 'omics', 'traditional']\n",
      "\n",
      "Cluster 242\n",
      "['study', 'previous']\n",
      "\n",
      "Cluster 243\n",
      "['consistent', 'metastatic', 'pc', 'penguin', 'cnas', 'western', 'xenograft', 'hnf1b', 'tp53']\n",
      "\n",
      "Cluster 244\n",
      "['number', 'small', 'copy', 'across']\n",
      "\n",
      "Cluster 245\n",
      "['increased', 'or', 'concentration', 'glucose', 'pressure', 'serum', 'fasting']\n",
      "\n",
      "Cluster 246\n",
      "['m']\n",
      "\n",
      "Cluster 247\n",
      "['codon', 'synonymous', 'usage']\n",
      "\n",
      "Cluster 248\n",
      "['software', 'r', 'version', 'package', 'implemented', 'suite']\n",
      "\n",
      "Cluster 249\n",
      "['expression', 'profile']\n",
      "\n",
      "Cluster 250\n",
      "['promoter', 'signal', 'ctcf', 'peg3', 'enhancer', 'cohesin']\n",
      "\n",
      "Cluster 251\n",
      "['panel', 'genotyping', 'imputation', 'hapmap', 'genotyped', 'snvs']\n",
      "\n",
      "Cluster 252\n",
      "['unique', 'core', 'etc', 'encoded', 'involves', 'conservation', 'transposon', 'conversion', 'rich']\n",
      "\n",
      "Cluster 253\n",
      "['family', 'activity', 'like', 'receptor', 'member', 'kinase', 'membrane', 'antigen', 'p53', 'channel']\n",
      "\n",
      "Cluster 254\n",
      "['four', 'corresponding', 'six', 'left', 'eight', 'median', 'plotted', 'versus', 'leaf', 'oligonucleotides', '1c']\n",
      "\n",
      "Cluster 255\n",
      "['diabetes', 'syndrome', 'obesity', 'chronic', 'poor', 'cardiovascular', 'symptom', 'coronary']\n",
      "\n",
      "Cluster 256\n",
      "['crocodylus', 'crocodilian', 'probably', 'karyotype', 'earlier']\n",
      "\n",
      "Cluster 257\n",
      "['energy', 'gram']\n",
      "\n",
      "Cluster 258\n",
      "['difference', 'similar', 'subpopulation']\n",
      "\n",
      "Cluster 259\n",
      "['comparative', 'thaliana', 'successfully', 'arabidopsis', 'sqtls']\n",
      "\n",
      "Cluster 260\n",
      "['free', 'water', 'tank']\n",
      "\n",
      "Cluster 261\n",
      "['different']\n",
      "\n",
      "Cluster 262\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['primate', 'eukaryotic', 'throughout', 'transposable', 'numt']\n",
      "\n",
      "Cluster 263\n",
      "['see', 'summarized', 'listed']\n",
      "\n",
      "Cluster 264\n",
      "['proposed']\n",
      "\n",
      "Cluster 265\n",
      "['for']\n",
      "\n",
      "Cluster 266\n",
      "['list', 'deg']\n",
      "\n",
      "Cluster 267\n",
      "['rt', 'iii', 'transgenic']\n",
      "\n",
      "Cluster 268\n",
      "['9', '15', '16', '12', '11', '13', '14']\n",
      "\n",
      "Cluster 269\n",
      "['model', 'prediction']\n",
      "\n",
      "Cluster 270\n",
      "['sequence', 'short', 'longer', 'chromhmm']\n",
      "\n",
      "Cluster 271\n",
      "['contains', 'content', 'genus', 'occurrence', 'representing', 'respect', 'symbol', 'match']\n",
      "\n",
      "Cluster 272\n",
      "['complex']\n",
      "\n",
      "Cluster 273\n",
      "['moreover', 'generally', 'gain', 'majority', 'seems', 'determining', 'timing', 'occurring']\n",
      "\n",
      "Cluster 274\n",
      "['component']\n",
      "\n",
      "Cluster 275\n",
      "['provided', 'public', 'currently', 'biomedical', 'publication', 'repository', 'pubmed', 'guideline']\n",
      "\n",
      "Cluster 276\n",
      "['rna']\n",
      "\n",
      "Cluster 277\n",
      "['the']\n",
      "\n",
      "Cluster 278\n",
      "['wide']\n",
      "\n",
      "Cluster 279\n",
      "['system']\n",
      "\n",
      "Cluster 280\n",
      "['ethnic', 'validated', 'asia', 'replicated', 'parent', 'trend', 'pulse']\n",
      "\n",
      "Cluster 281\n",
      "['better', 'improve', 'investigate', 'evaluate', 'validate', 'predicting']\n",
      "\n",
      "Cluster 282\n",
      "['peptide']\n",
      "\n",
      "Cluster 283\n",
      "['non', 'coding', 'transcript']\n",
      "\n",
      "Cluster 284\n",
      "['sirna', 'antibody', 'casz1']\n",
      "\n",
      "Cluster 285\n",
      "['in']\n",
      "\n",
      "Cluster 286\n",
      "['site', 'recombination', 'near', 'splice']\n",
      "\n",
      "Cluster 287\n",
      "['text', 'bioinformatics', 'service', 'source', 'processing', 'access', 'corpus', 'language', 'authorization', 'bionlp']\n",
      "\n",
      "Cluster 288\n",
      "['individual', 'status', 'gender']\n",
      "\n",
      "Cluster 289\n",
      "['collected', 'tcga']\n",
      "\n",
      "Cluster 290\n",
      "['red', 'color', 'beta', 'blue', 'assigned', 'green', 'box', 'dot']\n",
      "\n",
      "Cluster 291\n",
      "['line']\n",
      "\n",
      "Cluster 292\n",
      "['analysis']\n",
      "\n",
      "Cluster 293\n",
      "['ranked']\n",
      "\n",
      "Cluster 294\n",
      "['dna']\n",
      "\n",
      "Cluster 295\n",
      "['extracted', 'requires', 'collection', 'stored', 'broad', 'commercial']\n",
      "\n",
      "Cluster 296\n",
      "['regulatory', 'alternative', 'splicing', 'hotair']\n",
      "\n",
      "Cluster 297\n",
      "['seven', 'consisting', 'appeared', '63', 'km']\n",
      "\n",
      "Cluster 298\n",
      "['diagnostic', 'despite', 'aml', 'already', 'clinically']\n",
      "\n",
      "Cluster 299\n",
      "['et']\n",
      "\n",
      "Cluster 300\n",
      "['alteration', 'evolution', 'organization', 'instability', 'epigenomic']\n",
      "\n",
      "Cluster 301\n",
      "['signaling', 'cycle', 'dependent', 'nf', 'kdo', 'ifn', 'inhibition', 'activated', 'br', 'tnf', 'stimulation']\n",
      "\n",
      "Cluster 302\n",
      "['size', 'larger']\n",
      "\n",
      "Cluster 303\n",
      "['identify', 'provide', 'discovery', 'identification', 'possible', 'useful', 'find', 'identifying', 'interest', 'discover']\n",
      "\n",
      "Cluster 304\n",
      "['pathway']\n",
      "\n",
      "Cluster 305\n",
      "['set']\n",
      "\n",
      "Cluster 306\n",
      "['japanese', 'european', 'chinese']\n",
      "\n",
      "Cluster 307\n",
      "['kegg', 'kyoto']\n",
      "\n",
      "Cluster 308\n",
      "['algorithm', 'finally', 'constructed', 'then', 'filtering', 'fixed', 'training', 'filter', 'applying', 'partition', 'gatk']\n",
      "\n",
      "Cluster 309\n",
      "['designed', 'memory', 'full', 'window', 'built', 'converted', 'script', 'bam', 'java']\n",
      "\n",
      "Cluster 310\n",
      "['affect', 'likely', 'influence', 'affected', 'hearing', 'affecting']\n",
      "\n",
      "Cluster 311\n",
      "['although', 'finding', 'evidence', 'suggest']\n",
      "\n",
      "Cluster 312\n",
      "['matched']\n",
      "\n",
      "Cluster 313\n",
      "['required', 'especially', 'direct', 'become', 'microbiome', 'complexity', 'particularly', 'numerous', 'aspect', 'question', 'characterization', 'discus', 'opportunity']\n",
      "\n",
      "Cluster 314\n",
      "['considered']\n",
      "\n",
      "Cluster 315\n",
      "['highly']\n",
      "\n",
      "Cluster 316\n",
      "['locus', 'mtdna', 'polymorphic']\n",
      "\n",
      "Cluster 317\n",
      "['atc', 'sensitive', 'sjs', 'chromothripsis', 'acute', 'resistant', 'leukemia', 'solid', 'organ', 'thyroid', 'circulating']\n",
      "\n",
      "Cluster 318\n",
      "['food', 'central', 'introduction']\n",
      "\n",
      "Cluster 319\n",
      "['virus', 'viral', 'host', 'bacterial']\n",
      "\n",
      "Cluster 320\n",
      "['role', 'play', 'suggested', 'lead', 'contribute', 'pathogenesis', 'tumorigenesis', 'crucial']\n",
      "\n",
      "Cluster 321\n",
      "['nc', 'healthy', 'young', 'diagnosed', 'mortality']\n",
      "\n",
      "Cluster 322\n",
      "['identified']\n",
      "\n",
      "Cluster 323\n",
      "['environment', 'introduced', 'comprehensive', 'metagenomics', 'advanced', 'making', 'allow', 'extensive', 'suitable', 'enables', 'facilitate', 'principle', 'propose', 'attempt', 'methodology', 'describe', 'adopted', 'scientist', 'enable', 'perspective']\n",
      "\n",
      "Cluster 324\n",
      "['category']\n",
      "\n",
      "Cluster 325\n",
      "['selected', 'top']\n",
      "\n",
      "Cluster 326\n",
      "['detection', 'targeted', 'cost', 'wes', 'oncochase']\n",
      "\n",
      "Cluster 327\n",
      "['many', 'several', 'novel', 'recently', 'discovered']\n",
      "\n",
      "Cluster 328\n",
      "['analyzed']\n",
      "\n",
      "Cluster 329\n",
      "['selection', 'meta', 'estimation', 'fst', 'estimating']\n",
      "\n",
      "Cluster 330\n",
      "['article', 'paper']\n",
      "\n",
      "Cluster 331\n",
      "['characteristic']\n",
      "\n",
      "Cluster 332\n",
      "['module', 'string', 'hypothetical']\n",
      "\n",
      "Cluster 333\n",
      "['called', 'by', 'local', 'construct', 'correct', 'taxonomic', 'consensus', 'msdna']\n",
      "\n",
      "Cluster 334\n",
      "['third', 'both', 'no', 'flow', 'repeated', 'separate', 'accepted', 'consisted', 'remained']\n",
      "\n",
      "Cluster 335\n",
      "['position']\n",
      "\n",
      "Cluster 336\n",
      "['variable', 'variance', 'linear', 'statistic', 'likelihood', 'weighted', 'en', 'slr']\n",
      "\n",
      "Cluster 337\n",
      "['000', 'million', 'approximately', '500']\n",
      "\n",
      "Cluster 338\n",
      "['reference', 'mapping', 'calling']\n",
      "\n",
      "Cluster 339\n",
      "['further', 'taken', 'supported', 'hence', 'neuropeptides', 'upon', 'mode', 'towards']\n",
      "\n",
      "Cluster 340\n",
      "['known', 'major', 'linked', 'responsible', 'causing', 'thought', 'gut', 'pathogenic', 'functionally']\n",
      "\n",
      "Cluster 341\n",
      "['error', 'comparing', 'background', 'sampling', 'achieve']\n",
      "\n",
      "Cluster 342\n",
      "['main', 'good', 'basis', 'evaluation', 'improved', 'assessment', 'predictor', 'decision']\n",
      "\n",
      "Cluster 343\n",
      "['biomarker', 'deps']\n",
      "\n",
      "Cluster 344\n",
      "['residue', 'bond', 'hydrogen', 'rpob', 'formed', 'kir6']\n",
      "\n",
      "Cluster 345\n",
      "['measurement', 'correction', 'removing', 'rank', 'principal', 'classifier', 'computed', 'mixed', 'sum']\n",
      "\n",
      "Cluster 346\n",
      "['also']\n",
      "\n",
      "Cluster 347\n",
      "['various', 'diverse']\n",
      "\n",
      "Cluster 348\n",
      "['mutated']\n",
      "\n",
      "Cluster 349\n",
      "['nuclear', 'co', 'downstream', 'precursor', 'subunit', 'encoding', 'hp1', 'encodes', 'phosphorylation', 'mature', 'nucleus']\n",
      "\n",
      "Cluster 350\n",
      "['program', 'web', 'server', 'page']\n",
      "\n",
      "Cluster 351\n",
      "['enzyme', 'via', 'synthesis', 'repair', 'damage']\n",
      "\n",
      "Cluster 352\n",
      "['extended', 'account', 'hard', 'special', 'briefly', 'actual']\n",
      "\n",
      "Cluster 353\n",
      "['distributed', 'largest', 'divergence', 'ancestral']\n",
      "\n",
      "Cluster 354\n",
      "['combination']\n",
      "\n",
      "Cluster 355\n",
      "['drug']\n",
      "\n",
      "Cluster 356\n",
      "['genome']\n",
      "\n",
      "Cluster 357\n",
      "['95', 'h', 'min', '2017', 'day', '72', '82', '120', 'temperature', '94', '91', 'ml']\n",
      "\n",
      "Cluster 358\n",
      "['score', 'matrix', 'parameter', 'weight', 'random', 'z', 'probability']\n",
      "\n",
      "Cluster 359\n",
      "['highest', 'lowest', 'mol', 'kcal']\n",
      "\n",
      "Cluster 360\n",
      "['author']\n",
      "\n",
      "Cluster 361\n",
      "['without', 'determine', 'because', 'difficult', 'cannot', 'enough', 'sufficient']\n",
      "\n",
      "Cluster 362\n",
      "['cohort', 'participant', 'ansan', 't2d', '842']\n",
      "\n",
      "Cluster 363\n",
      "['quantitative', 'bias', 'measuring']\n",
      "\n",
      "Cluster 364\n",
      "['method']\n",
      "\n",
      "Cluster 365\n",
      "['factor']\n",
      "\n",
      "Cluster 366\n",
      "['hit', 'determined', 'angle', 'atom', 'ramachandran']\n",
      "\n",
      "Cluster 367\n",
      "['time']\n",
      "\n",
      "Cluster 368\n",
      "['show']\n",
      "\n",
      "Cluster 369\n",
      "['order']\n",
      "\n",
      "Cluster 370\n",
      "['so', 'spectrum', 'extract', 'capacity', 'adaptation', 'consequently', 'regarded']\n",
      "\n",
      "Cluster 371\n",
      "['gov', 'nih', 'retrieved', 'nlm', 'online', 'uk', 'uniprot']\n",
      "\n",
      "Cluster 372\n",
      "['control']\n",
      "\n",
      "Cluster 373\n",
      "['cgi']\n",
      "\n",
      "Cluster 374\n",
      "['we']\n",
      "\n",
      "Cluster 375\n",
      "['kind', 'context', 'mentioned', 'helpful', 'besides', 'demonstrate', 'offer', 'fully', 'explored']\n",
      "\n",
      "Cluster 376\n",
      "['treated', 'hdfs', 'dmso']\n",
      "\n",
      "Cluster 377\n",
      "['culture', 'medium']\n",
      "\n",
      "Cluster 378\n",
      "['based', 'support', 'classification', 'way', 'concept', 'graph', 'machine', 'learning', 'semantic', 'attribute']\n",
      "\n",
      "Cluster 379\n",
      "['replication']\n",
      "\n",
      "Cluster 380\n",
      "['indels']\n",
      "\n",
      "Cluster 381\n",
      "['mitochondrial']\n",
      "\n",
      "Cluster 382\n",
      "['browser', 'ucsc']\n",
      "\n",
      "Cluster 383\n",
      "['ii', 'stem']\n",
      "\n",
      "Cluster 384\n",
      "['environmental', 'studied', 'unknown', 'action', 'physiological', 'controlling', 'ro', 'understood']\n",
      "\n",
      "Cluster 385\n",
      "['lineage', 'abundant', 'subfamily', 'mammalian', 'mammal']\n",
      "\n",
      "Cluster 386\n",
      "['mir', '21', '3p']\n",
      "\n",
      "Cluster 387\n",
      "['index']\n",
      "\n",
      "Cluster 388\n",
      "['effective', 'decade', 'rapidly', 'more', 'rapid', 'successful']\n",
      "\n",
      "Cluster 389\n",
      "['trna']\n",
      "\n",
      "Cluster 390\n",
      "['point', 'name', 'sub', 'phylogenetic', 'contiguous', 'maximal']\n",
      "\n",
      "Cluster 391\n",
      "['screening']\n",
      "\n",
      "Cluster 392\n",
      "['genotype', 'five', 'criterion', 'imputed']\n",
      "\n",
      "Cluster 393\n",
      "['male', 'female', 'men', '69', 'oh']\n",
      "\n",
      "Cluster 394\n",
      "['rate']\n",
      "\n",
      "Cluster 395\n",
      "['found']\n",
      "\n",
      "Cluster 396\n",
      "['chemical', 'dynamic', 'recognition', 'ape1']\n",
      "\n",
      "Cluster 397\n",
      "['range']\n",
      "\n",
      "Cluster 398\n",
      "['development', 'treatment', 'early', 'therapy', 'therapeutic', 'prognosis']\n",
      "\n",
      "Cluster 399\n",
      "['evolutionary']\n",
      "\n",
      "Cluster 400\n",
      "['31', '32', '38', '39', '46', 'il', '43', '56', '44', '47', '49']\n",
      "\n",
      "Cluster 401\n",
      "['demonstrated', 'increasing', 'uterine', 'remains', 'microbiota']\n",
      "\n",
      "Cluster 402\n",
      "['additional', 'easily', 'must', 'combining', 'require', 'popular', 'get', 'taking', 'material', 'speed', 'apply', 'select', 'combine', 'meaningful', 'avoid', 'expect', 'quantity']\n",
      "\n",
      "Cluster 403\n",
      "['classified', 'distinct']\n",
      "\n",
      "Cluster 404\n",
      "['information']\n",
      "\n",
      "Cluster 405\n",
      "['methylation']\n",
      "\n",
      "Cluster 406\n",
      "['effect', 'risk']\n",
      "\n",
      "Cluster 407\n",
      "['leu', 'ala', 'phe', 'tyr', 'val']\n",
      "\n",
      "Cluster 408\n",
      "['compared', 'obese']\n",
      "\n",
      "Cluster 409\n",
      "['functional', 'relationship']\n",
      "\n",
      "Cluster 410\n",
      "['derived']\n",
      "\n",
      "Cluster 411\n",
      "['association']\n",
      "\n",
      "Cluster 412\n",
      "['would', 'need', 'u', 'interesting', 'lack', 'give', 'reduce', 'reason', 'great']\n",
      "\n",
      "Cluster 413\n",
      "['change']\n",
      "\n",
      "Cluster 414\n",
      "['year']\n",
      "\n",
      "Cluster 415\n",
      "['result']\n",
      "\n",
      "Cluster 416\n",
      "['clustering', 'hierarchical']\n",
      "\n",
      "Cluster 417\n",
      "['higher', 'le', 'lower', 'greater']\n",
      "\n",
      "Cluster 418\n",
      "['within', 'kb', 'utr']\n",
      "\n",
      "Cluster 419\n",
      "['er', 'docked', 'conformation', 'phosphoglucosamine', 'mutase']\n",
      "\n",
      "Cluster 420\n",
      "['efficiency']\n",
      "\n",
      "Cluster 421\n",
      "['observation', 'interestingly', 'exhibit', 'imprinted']\n",
      "\n",
      "Cluster 422\n",
      "['measured', 'ct']\n",
      "\n",
      "Cluster 423\n",
      "['used']\n",
      "\n",
      "Cluster 424\n",
      "['molecular', 'ligand', 'compound', 'docking']\n",
      "\n",
      "Cluster 425\n",
      "['protein', 'molecule', 'metal']\n",
      "\n",
      "Cluster 426\n",
      "['predict', 'silico', 'theory', 'tensor', 'artificial']\n",
      "\n",
      "Cluster 427\n",
      "['case']\n",
      "\n",
      "Cluster 428\n",
      "['either', 'enriched', 'whose', 'hypermethylated']\n",
      "\n",
      "Cluster 429\n",
      "['level', 'body', '27a', 'g64']\n",
      "\n",
      "Cluster 430\n",
      "['fusion', 'sequenced', 'yeast', 'rrna', 'sanger', 'fission', 'clone']\n",
      "\n",
      "Cluster 431\n",
      "['instance']\n",
      "\n",
      "Cluster 432\n",
      "['infection', 'resistance', 'pathogen']\n",
      "\n",
      "Cluster 433\n",
      "['polymerase', 'product', 'reverse']\n",
      "\n",
      "Cluster 434\n",
      "['cnv', 'cnvs']\n",
      "\n",
      "Cluster 435\n",
      "['feature']\n",
      "\n",
      "Cluster 436\n",
      "['categorized']\n",
      "\n",
      "Cluster 437\n",
      "['followed', 'gel', 'separated', '96', 'buffer', 'extension', 'electrophoresis', 'mix']\n"
     ]
    }
   ],
   "source": [
    "# word - cluster number matching\n",
    "word_centroid_map = dict(zip(gniwv.index2word, index))\n",
    "\n",
    "# Print whole clutering results\n",
    "for cluster in range(num_clusters) :\n",
    "    print('\\nCluster',cluster)\n",
    "    words = []\n",
    "    for k, c in word_centroid_map.items():\n",
    "        if c == cluster :\n",
    "            words.append(k)\n",
    "    print(words)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multidimensional Scaling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate distribution\n",
    "dist = 1-cosine_similarity(gniwv.vectors)\n",
    "\n",
    "# initialize MDS\n",
    "MDS()\n",
    "\n",
    "# scaling multidimensions\n",
    "mds = MDS(n_components=2, dissimilarity = \"precomputed\", random_state=1)\n",
    "pos = mds.fit_transform(dist)\n",
    "\n",
    "xs, ys = pos[:,0], pos[:,1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualization of Word vectors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set colors and names of clusters\n",
    "cluster_colors = {35: '#1b9e77', 36: '#d95f02', 37: '#7570b3', 38: '#e7298a', 39: '#66a61e'}\n",
    "cluster_names = {35: 'Cluster35', \n",
    "                 36: 'Cluster36', \n",
    "                 37: 'Cluster37', \n",
    "                 38: 'Cluster38', \n",
    "                 39: 'Cluster39'\n",
    "                }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "# Create (x-position(2D), y-position(2D), word, cluster number) table\n",
    "df_all = pd.DataFrame(dict(x=xs, y=ys,word=gniwv.index2word, label=index))\n",
    "\n",
    "# Show only five Clusters, 35, 36, 37, 38, 39\n",
    "condition = ( df_all['label'] >= 35 ) & ( df_all['label'] < 40 )\n",
    "df = df_all[condition]\n",
    "\n",
    "# grouping with cluster number\n",
    "groups = df.groupby('label')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(17, 9)) # set size\n",
    "\n",
    "ax.margins(0.05)\n",
    "\n",
    "for cluseter_n, group in groups:\n",
    "    ax.plot(group.x, group.y, marker='o', linestyle='', ms=12, \n",
    "                label=cluster_names[cluseter_n],\n",
    "            color=cluster_colors[cluseter_n], \n",
    "            mec='none')\n",
    "    ax.set_aspect('auto')\n",
    "    ax.tick_params(\\\n",
    "        axis= 'x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom='off',      # ticks along the bottom edge are off\n",
    "        top='off',         # ticks along the top edge are off\n",
    "        labelbottom='off')\n",
    "    ax.tick_params(\\\n",
    "        axis= 'y',         # changes apply to the y-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left='off',      # ticks along the bottom edge are off\n",
    "        top='off',         # ticks along the top edge are off\n",
    "        labelleft='off')\n",
    "\n",
    "ax.legend(numpoints=1)\n",
    "\n",
    "for i in range(len(df)):\n",
    "    ax.text(df.iloc[i]['x'],\n",
    "            df.iloc[i]['y'],\n",
    "            df.iloc[i]['word'], size=12)  \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find the cluster which include a specific word"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find cluster result with specipic keyword\n",
    "def find_cluster_with_key(key):\n",
    "    # if there is not key in vocab, do nothing\n",
    "    if not key in word_centroid_map.keys() :\n",
    "        print(\"That word is not in keys\")\n",
    "    # if there is key in vocab, print cluster result\n",
    "    else :  \n",
    "        cluster = word_centroid_map[key]\n",
    "        print('\\nCluster',cluster)\n",
    "        words = []\n",
    "        for k, c in word_centroid_map.items():\n",
    "            if c == cluster :\n",
    "                words.append(k)\n",
    "        print(words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key = input(\"============= Find Clustering ==============\\nEnter a keyword: \")\n",
    "find_cluster_with_key(key)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
